File: engine/aurora/power.py
"""Power Engine — deney BAŞLAMADAN önce gereken örneklem büyüklüğünü hesaplar.

Neden gerekli: n=5 ile d=1.8 gibi büyük bir gerçek etki bile Bonferroni sonrası anlamlı çıkmaz
(güç ~%30). Yetersiz güçle yürütülen bir çalışma "etki yok" sonucunu kanıt sanma hatasına yol açar.
Önkayıt bu yüzden hedef etki büyüklüğünü ve gücü içermelidir.
"""
from __future__ import annotations

import numpy as np
from scipy import stats as sps

DEFAULT_ALPHA = 0.05
DEFAULT_POWER = 0.80


def power_two_sample(n: int, effect_d: float, comparisons: int = 1, alpha: float = DEFAULT_ALPHA,
                     sims: int = 4000, seed: int = 0) -> float:
    """Welch t + Bonferroni düzeltmesi altında gerçek `effect_d` etkisini yakalama olasılığı.

    Simülasyon tabanlıdır: analitik formüller Welch df ve çoklu karşılaştırmayı birlikte
    kapsamadığı için, gerçekte kullanılan analiz hattının aynısı çalıştırılır.
    """
    if n < 2:
        return 0.0
    rng = np.random.default_rng(seed)
    a = rng.normal(effect_d, 1.0, (sims, n))
    b = rng.normal(0.0, 1.0, (sims, n))
    _, p = sps.ttest_ind(a, b, axis=1, equal_var=False)
    return float(np.mean(np.minimum(1.0, p * max(1, comparisons)) < alpha))


def required_n(effect_d: float, comparisons: int = 1, target_power: float = DEFAULT_POWER,
               alpha: float = DEFAULT_ALPHA, max_n: int = 400, sims: int = 2000) -> int | None:
    """Hedef güce ulaşmak için koşul başına gereken deneme sayısı."""
    for n in range(4, max_n + 1):
        if power_two_sample(n, effect_d, comparisons, alpha, sims) >= target_power:
            return n
    return None


def detectable_effect(n: int, comparisons: int = 1, target_power: float = DEFAULT_POWER,
                      alpha: float = DEFAULT_ALPHA, sims: int = 2000) -> float | None:
    """Verilen n ile hedef güçte saptanabilecek en küçük etki büyüklüğü (MDE)."""
    for d in np.arange(0.1, 4.01, 0.1):
        if power_two_sample(n, float(d), comparisons, alpha, sims) >= target_power:
            return round(float(d), 2)
    return None


def plan(conditions: int, effect_d: float = 0.8, target_power: float = DEFAULT_POWER,
         minutes_per_trial: float = 10.0, alpha: float = DEFAULT_ALPHA) -> dict:
    """Bir çalışma için tam örneklem planı: n, toplam deneme, süre, MDE tablosu."""
    if conditions < 2:
        raise ValueError("at least 2 conditions (including a control) are required")
    k = conditions - 1
    n = required_n(effect_d, k, target_power, alpha)
    total = n * conditions if n else None
    return {
        "conditions": conditions,
        "comparisons": k,
        "alpha": alpha,
        "correction": "bonferroni",
        "target_effect_d": effect_d,
        "target_power": target_power,
        "n_per_condition": n,
        "total_trials": total,
        "total_hours": round(total * minutes_per_trial / 60, 1) if total else None,
        "note": "An underpowered study cannot support a 'no effect' conclusion. "
                "Fix n before collecting data; report the minimum detectable effect either way.",
    }


def power_table(ns: list[int], ds: list[float], comparisons: int = 3,
                alpha: float = DEFAULT_ALPHA) -> list[dict]:
    """n × d güç tablosu (önkayıt belgesine eklenir)."""
    return [{"n_per_condition": n,
             **{f"d={d}": round(power_two_sample(n, d, comparisons, alpha), 3) for d in ds}}
            for n in ns]

File: app/src/core/power.ts
/** Power — deney başlamadan önce güç/örneklem hesabı (engine/aurora/power.py eşleniği).
 *
 * Neden: n=5 ile d=1.8 gibi büyük bir gerçek etki bile Bonferroni sonrası anlamlı çıkmaz.
 * Yetersiz güçle yürütülen bir çalışma "etki yok" sonucunu desteklemez.
 */
import { tCdf } from "./stats";

/** Box–Muller ile deterministik normal örnek (tohumlu). */
const mulberry = (seed: number) => () => {
  let t = (seed += 0x6d2b79f5);
  t = Math.imul(t ^ (t >>> 15), t | 1);
  t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
  return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
};
const normal = (rng: () => number, mean: number) => {
  const u = Math.max(rng(), 1e-12);
  return mean + Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * rng());
};

const meanOf = (x: number[]) => x.reduce((a, b) => a + b, 0) / x.length;
const varOf = (x: number[]) => {
  const m = meanOf(x);
  return x.reduce((a, b) => a + (b - m) ** 2, 0) / (x.length - 1);
};

/** Welch t + Bonferroni altında gerçek etkiyi yakalama olasılığı (simülasyon). */
export const powerTwoSample = (n: number, effectD: number, comparisons = 1, alpha = 0.05, sims = 600, seed = 0): number => {
  if (n < 2) return 0;
  const rng = mulberry(seed);
  let hits = 0;
  for (let s = 0; s < sims; s++) {
    const a = Array.from({ length: n }, () => normal(rng, effectD));
    const b = Array.from({ length: n }, () => normal(rng, 0));
    const va = varOf(a) / n, vb = varOf(b) / n;
    const t = (meanOf(a) - meanOf(b)) / Math.sqrt(va + vb);
    const df = (va + vb) ** 2 / (va ** 2 / (n - 1) + vb ** 2 / (n - 1));
    const p = 2 * (1 - tCdf(Math.abs(t), df));
    if (Math.min(1, p * Math.max(1, comparisons)) < alpha) hits++;
  }
  return hits / sims;
};

/** Hedef güç için koşul başına gereken deneme sayısı. */
export const requiredN = (effectD: number, comparisons = 1, targetPower = 0.8, alpha = 0.05): number | null => {
  for (let n = 4; n <= 200; n += n < 40 ? 1 : 5) {
    if (powerTwoSample(n, effectD, comparisons, alpha, 400, n) >= targetPower) return n;
  }
  return null;
};

/** Verilen n ile saptanabilecek en küçük etki (MDE). */
export const detectableEffect = (n: number, comparisons = 1, targetPower = 0.8, alpha = 0.05): number | null => {
  for (let d = 0.2; d <= 4.0; d += 0.1) {
    if (powerTwoSample(n, d, comparisons, alpha, 400, Math.round(d * 10)) >= targetPower) return Math.round(d * 10) / 10;
  }
  return null;
};

export interface PowerPlan {
  conditions: number;
  comparisons: number;
  nPerCondition: number;
  totalTrials: number;
  totalHours: number;
  detectableD: number | null;
  powerAtTarget: number;
  adequate: boolean;
}

/** Kullanıcının seçtiği tasarımın gücünü değerlendirir. */
export const evaluateDesign = (conditions: number, repetitions: number, minutesPerTrial: number, targetD = 0.8): PowerPlan => {
  const comparisons = Math.max(1, conditions - 1);
  const powerAtTarget = powerTwoSample(repetitions, targetD, comparisons, 0.05, 400, 11);
  return {
    conditions,
    comparisons,
    nPerCondition: repetitions,
    totalTrials: conditions * repetitions,
    totalHours: Math.round((conditions * repetitions * minutesPerTrial) / 6) / 10,
    detectableD: detectableEffect(repetitions, comparisons),
    powerAtTarget,
    adequate: powerAtTarget >= 0.8,
  };
};
